mock/py/mockbuild/plugins/rpmautospec.py

```diff
@@ -0,0 +1,117 @@
+
+from mockbuild.exception import ConfigError
+from mockbuild.trace_decorator import getLog, traceLog
+
+requires_api_version = "1.1"
+
+
+@traceLog()
+def init(plugins, conf, buildroot):
+    """Register the rpmautospec plugin with mock."""
+    RpmautospecPlugin(plugins, conf, buildroot)
+
+
+class RpmautospecPlugin:
+    """Fill in release and changelog from git history using rpmautospec"""
+
+    @traceLog()
+    def __init__(self, plugins, conf, buildroot):
+        self.buildroot = buildroot
+        self.config = buildroot.config
+        self.opts = conf
+        self.log = getLog()
+
+        if "cmd_base" not in self.opts:
+            raise ConfigError("The 'rpmautospec_opts.cmd_base' is unset")
+
+        plugins.add_hook("pre_srpm_build", self.attempt_process_distgit)
+        self.log.info("rpmautospec: initialized")
+
+    @traceLog()
+    def attempt_process_distgit(
+        self,
+        host_chroot_spec: Union[Path, str],
+        host_chroot_sources: Optional[Union[Path, str]],
+    ) -> None:
+        """Attempt to process a spec file with rpmautospec."""
+        # Set up variables and check prerequisites.
+        if not host_chroot_sources:
+            self.log.debug("Sources not specified, skipping rpmautospec preprocessing.")
+            return
+
+        host_chroot_spec = Path(host_chroot_spec)
+        host_chroot_sources = Path(host_chroot_sources)
+        if not host_chroot_sources.is_dir():
+            self.log.debug(
+                "Sources not a directory, skipping rpmautospec preprocessing."
+            )
+            return
+
+        distgit_git_dir = host_chroot_sources / ".git"
+        if not distgit_git_dir.is_dir():
+            self.log.debug(
+                "Sources is not a git repository, skipping rpmautospec preprocessing."
+            )
+            return
+
+        host_chroot_sources_spec = host_chroot_sources / host_chroot_spec.name
+        if not host_chroot_sources_spec.is_file():
+            self.log.debug(
+                "Sources doesn’t contain spec file, skipping rpmautospec preprocessing."
+            )
+            return
+
+        with host_chroot_spec.open("rb") as spec, host_chroot_sources_spec.open(
+            "rb"
+        ) as sources_spec:
+            if spec.read() != sources_spec.read():
+                self.log.warning(
+                    "Spec file inside and outside sources are different, skipping rpmautospec"
+                    " preprocessing."
+                )
+                return
+
+        if not specfile_uses_rpmautospec(host_chroot_sources_spec):
+            self.log.debug(
+                "Spec file doesn’t use rpmautospec, skipping rpmautospec preprocessing."
+            )
+            return
+
+        # Install the `rpmautospec` command line tool into the build root.
+        if self.opts.get("requires", None):
+            self.buildroot.pkg_manager.install_as_root(*self.opts["requires"], check=True)
```

Comment:
The question still applies..  Depending on the packaged-in-chroot rpmautospec means that the behavior may vary depending on the Mock --chroot option.

Reply:
Sorry missed this comment of yours so far. What do you mean? I’ve looked up the `--chroot` option but nothing jumps at me how this is related to using rpmautospec as packaged in the chroot.

Reply:
That's the point of `self.opts.get("requires", None)`, right?  You want to install some packages into the build root.  Or am I missing something?

Reply:
OK, I think I can translate:

@praiskup Are you asking "What happens if the repos that the mock chroot is using don't contain the packages this is requesting?"

That will throw an exception here, if I'm reading things right. Maybe we should catch it and make sure there's a reasonable error message written, but if the specfile requires rpmautospec and rpmautospec isn't present in the targeted buildroot, failing is the right outcome.

Reply:
The default assume rpmautospec and dependencies need to be installed, but we want to have the option available to just put it in the image (and then skip attempts to install it). Or maybe I’m still not understanding the question. 😉

Reply:
Thank you for the comments above.

The question I wanted to ask is whether we need to depend on both (a) Autospec on host, as the hard Requires dependency, and (b) in-chroot, installed using `self.opts.get("requires", None)`.

Can't we use host's `Autospec` for everything?

Reply:
Said that, I'd like to clear that question out -- but I agree that, if we have to fail for missing in-chroot package, a nice error message is worth it.

Reply:
I assume, that another exception should be thrown so mock can signal a problem to the caller (e.g. Koji), right?